tests/unit/recommenders/models/test_ncf_dataset.py

```diff
@@ -1,120 +1,190 @@
+    with pytest.raises(Exception):
+        datafile = DataFile(
+            train_path, DEFAULT_USER_COL, DEFAULT_ITEM_COL, DEFAULT_RATING_COL, col_test_batch=None, binary=True
+        )
+
+
+@pytest.mark.gpu
+def test_negative_sampler(caplog):
+    user = 1
+    n_samples = 3
+    user_positive_item_pool = {1, 2}
+    item_pool = {1, 2, 3, 4, 5}
+    sample_with_replacement = False
+    sampler = NegativeSampler(user, n_samples, user_positive_item_pool, item_pool, sample_with_replacement)
+    assert sampler.n_samples == 3
+    samples = sampler.sample()
+    assert set(samples) == item_pool.difference(user_positive_item_pool)
+
+    # test sampler adjusts n_samples down if population is too small and that it raises a warning
+    n_samples = 4
+    sampler = NegativeSampler(user, n_samples, user_positive_item_pool, item_pool, sample_with_replacement)
+    assert sampler.n_samples == 3
+    assert "The population of negative items to sample from is too small for user 1" in caplog.text
+
+    # test sampling with replacement returns requested number of samples despite small population
+    sampler = NegativeSampler(user, n_samples, user_positive_item_pool, item_pool, sample_with_replacement=True)
+    assert sampler.n_samples == 4
+    assert len(sampler.sample()) == n_samples
+
+
+@pytest.mark.gpu
+def test_train_loader(tmp_path, dataset_ncf_files_sorted):
+    train_path, _, _ = dataset_ncf_files_sorted
+    train = pd.read_csv(train_path)
+    users = train[DEFAULT_USER_COL].unique()
+    items = train[DEFAULT_ITEM_COL].unique()
+
+    n_neg = 1
+    dataset = Dataset(train_path, n_neg=n_neg)
+    assert dataset.n_users == len(users)
+    assert dataset.n_items == len(items)
+    assert set(dataset.user2id.keys()) == set(users)
+    assert set(dataset.item2id.keys()) == set(items)
+    assert len(set(dataset.user2id.values())) == len(users)
+    assert len(set(dataset.item2id.values())) == len(items)
+
+    # test number of batches and data size is as expected after loading all training data
+    full_data_len = train.shape[0] * 2
+    batch_size = full_data_len // 10
+    expected_batches = full_data_len // batch_size
+    train_save_path = os.path.join(tmp_path, "train_full.csv")
+    batch_records = []
+    for batch in dataset.train_loader(batch_size, shuffle_size=batch_size, yield_id=True, write_to=train_save_path):
+        assert type(batch[0][0]) == int
+        assert type(batch[1][0]) == int
+        assert type(batch[2][0]) == float
+        batch_data = {
+            DEFAULT_USER_COL: [dataset.id2user[user] for user in batch[0]],
+            DEFAULT_ITEM_COL: [dataset.id2item[item] for item in batch[1]],
+            DEFAULT_RATING_COL: batch[2]
+        }
+        batch_records.append(pd.DataFrame(batch_data))
+    
+    assert len(batch_records) == expected_batches
+    train_loader_df = pd.concat(batch_records).reset_index(drop=True)
+    assert train_loader_df.shape[0] == expected_batches * batch_size
```

Comment:
Isn't it possible that the last batch will be smaller than batch_size and hence that this assertion is false?